rasa/core/nlg/generator.py

```diff
@@ -91,3 +90,121 @@ def _load_from_module_name_in_endpoint_config(
+        self,
+        utter_action: Text,
+        output_channel: Text,
+        filled_slots: Dict[Text, Any],
+    ) -> List[Dict[Text, Any]]:
+        """Returns array of responses that fit the channel, action and condition."""
+        default_responses = list(
+            filter(
+                lambda x: (x.get(RESPONSE_CONDITION) is None),
+                self.responses[utter_action],
+            )
+        )
+        conditional_responses = list(
+            filter(
+                lambda x: (
+                    x.get(RESPONSE_CONDITION)
+                    and self._matches_filled_slots(
+                        filled_slots=filled_slots, response=x
+                    )
+                ),
+                self.responses[utter_action],
+            )
+        )
+
+        conditional_channel = list(
+            filter(lambda x: (x.get(CHANNEL) == output_channel), conditional_responses)
+        )
+        conditional_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), conditional_responses)
+        )
+        default_channel = list(
+            filter(lambda x: (x.get(CHANNEL) == output_channel), default_responses)
+        )
+        default_no_channel = list(
+            filter(lambda x: (x.get(CHANNEL) is None), default_responses)
+        )
+
+        if conditional_channel:
+            return conditional_channel
+
+        if default_channel:
+            return default_channel
+
+        if conditional_no_channel:
+            return conditional_no_channel
+
+        return default_no_channel
+
+    def get_response_variation_id(
+        self,
+        utter_action: Text,
+        tracker: DialogueStateTracker,
+        output_channel: Text,
+    ) -> Optional[Text]:
+        """Returns the first matched response variation ID that fits the channel, action and condition."""  # noqa: E501
+        filled_slots = tracker.current_slot_values()
+        if utter_action in self.responses:
+            eligible_variations = self.responses_for_utter_action(
+                utter_action, output_channel, filled_slots
+            )
+            response_ids_are_valid = self._validate_response_ids(eligible_variations)
+
+            if eligible_variations and response_ids_are_valid:
+                return eligible_variations[0].get("id")
+
+        return None
+
+    @staticmethod
+    def _validate_response_ids(response_variations: List[Dict[Text, Any]]) -> bool:
+        """Checks that the response IDs of a particular utter_action are unique.
+
+        Args:
+            response_variations: The response variations to validate.
+
+        Returns:
+            True if the response IDs are unique, False otherwise.
+        """
+        response_ids = set()
+        for response_variation in response_variations:
+            response_variation_id = response_variation.get("id")
+            if response_variation_id:
+                if response_variation_id in response_ids:
```

Comment:
I think the condition `if response_variation_id in response_ids:` is sufficient and also checks for empty variable, what do you think?

Reply:
I'd combine the two actually by an AND to remove the indentation, but the first check is still required because we could be adding None values to the set in L208.

Reply:
Addressed: https://github.com/RasaHQ/rasa/pull/12575/commits/27e98159b985b91c4a2204fc18cc089e54d75efb